File: app/src/main/java/com/example/domain/usecase/LoginUseCase.kt
package com.example.domain.usecase

import com.example.domain.model.AuthChallenge
import com.example.domain.repo.AuthRepository

/**
 * Orchestrates login actions and session handshakes for the presentation layer.
 */
class LoginUseCase(private val authRepository: AuthRepository) {

    /**
     * Prepares authentication challenge properties to kickstart Custom Tabs display.
     */
    suspend fun beginLogin(): AuthChallenge {
        return authRepository.beginLogin()
    }

    /**
     * Executes official endpoint code resolution handshake.
     */
    suspend fun completeLogin(code: String, codeVerifier: String): Result<Unit> {
        return authRepository.completeLogin(code, codeVerifier)
    }
}

File: ios/Curio/Curio/Domain/LoginUseCase.swift
import Foundation

/// Orchestrates login actions and session handshakes for the presentation layer. Ports
/// `class LoginUseCase` from `domain/usecase/LoginUseCase.kt`.
///
/// A stateless pass-through to `AuthRepository` (kept trivial for easy mocking — CONVENTIONS §12).
/// `Sendable` since it holds only a `Sendable` repository reference.
struct LoginUseCase: Sendable {
    private let authRepository: AuthRepository

    init(authRepository: AuthRepository) {
        self.authRepository = authRepository
    }

    /// Prepares the authentication challenge to kickstart the login sheet.
    func beginLogin() async throws -> AuthChallenge {
        try await authRepository.beginLogin()
    }

    /// Executes the official endpoint code-resolution handshake. Throws on failure (collapses the
    /// Kotlin `Result<Unit>`).
    func completeLogin(code: String, codeVerifier: String) async throws {
        try await authRepository.completeLogin(code: code, codeVerifier: codeVerifier)
    }
}
